Split global schema lock

MishaDemianenko opened the pull request:
Split global schema lock into independent label and relationship type locks
that will be used depending from type of index, constraint etc.
Completely remove SCHEMA resource from type of resources that can be locked.

Slaves in ha now will take labels and rel type locks on master exactly like
for nodes and relationship locks.

This commit performs split in kernel/ha code.
Cypher code not updated as part of this commit.

MishaDemianenko committed:
Split global schema lock

Split global schema lock into independent label and relationship type locks
that will be used depending from type of index, constraint etc.
Completely remove SCHEMA resource from type of resources that can be locked.

Slaves in ha now will take labels and rel type locks on master exactly like
for nodes and relationship locks.

This commit performs split in kernel/ha code.
Cypher code not updated as part of this commit.

MishaDemianenko committed:
Do not allow construction of index descriptor for non valid
label id, property id, relationship type id.

MishaDemianenko committed:
Simplify DatabaseSchemaState hierarchy, collect label ids during planning.

MishaDemianenko committed:
Take shared label locks for planed cypher queries.
Release then in case if there was concurrent schema state clear and retry.
Fix failing tests

MishaDemianenko committed:
We no longer allow slaves to ignore shared schema lock, label locks are taken on master
and test is no longer relevant

MishaDemianenko committed:
Lock all the labels in case if cypher version is before 3.2

MishaDemianenko committed:
Remove not used method in test and not used deprecated method anymore.

MishaDemianenko committed:
Cleanup and javadocs

MishaDemianenko committed:
Lock and release label locks for query plans in bulk.

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/LockingStatementOperations.java:
```diff
@@ -481,34 +459,33 @@ public Value graphRemoveProperty( KernelStatement state, int propertyKeyId )
 
     @Override
-    public void acquireExclusive( KernelStatement state, ResourceType resourceType, long resourceId )
+    public void acquireExclusive( KernelStatement state, ResourceType resourceType, long... resourceId )
```
`resourceIds`? Similar for other methods below in that case.

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:
```diff
@@ -994,28 +1000,28 @@ public void uniqueIndexDrop( IndexDescriptor descriptor ) throws DropIndexFailur
     // <Locking>
     @Override
-    public void acquireExclusive( ResourceType type, long id )
+    public void acquireExclusive( ResourceType type, long... id )
```
could be `ids`?

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/ResourceTypes.java:
```diff
@@ -37,9 +37,11 @@
     RELATIONSHIP( 1, LockWaitStrategies.INCREMENTAL_BACKOFF ),
     GRAPH_PROPS( 2, LockWaitStrategies.INCREMENTAL_BACKOFF ),
-    SCHEMA( 3, LockWaitStrategies.INCREMENTAL_BACKOFF ),
+    // SCHEMA resource type had typeId 3 - skip it to avoid resource types conflicts
```
Good call!

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/community/CommunityLockClient.java:
```diff
@@ -270,21 +270,24 @@ private boolean reEnter( PrimitiveLongObjectMap<LockResource> localLocks, long r
+                LockResource resource = localLocks.get( resourceId );
+                if ( resource.releaseReference() != 0 )
+                {
+                    return;
```
Do we not want to try releasing any remaining resources after the first failure? I don't know, maybe this will terminate the transaction and release everything anyway, just noting this observation.

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/community/CommunityLockClient.java:
```diff
@@ -293,20 +296,23 @@ public void releaseShared( ResourceType resourceType, long resourceId )
+                LockResource resource = localLocks.get( resourceId );
+                if ( resource.releaseReference() != 0 )
+                {
+                    return;
```
See return comment above.

fickludd reviewed (approved):
The idea of splitting the schema lock is really good, let's hope that we can back that by some SOAK-run or something! A workload with regular reads and increased amounts of index/constraint changes should definitely benefit.

I think the cypher part of this look good, but it did not go through all the kernel parts. Some very minor naming comments, but nothing that has to be fixed.

chrisvest commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -586,11 +586,6 @@
-    @Description( "Whether or not to release the exclusive schema lock is while building uniqueness constraints index" )
-    @Internal
-    public static final Setting<Boolean> release_schema_lock_while_building_constraint = setting(
-            "unsupported.dbms.schema.release_lock_while_building_constraint", BOOLEAN, FALSE );
```
I don't think we can technically remove this. But we can certainly deprecate and ignore it.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java:
```diff
@@ -276,17 +276,21 @@ IndexDescriptor indexGetForSchema( LabelSchemaDescriptor descriptor )
 
-    void releaseExclusive( ResourceType type, long id );
-    void releaseShared(    ResourceType type, long id );
+    void releaseShared( ResourceType type, long... id );
```
We can pluralise `id`

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/LockingStatementOperations.java:
```diff
@@ -98,18 +94,7 @@ public LockingStatementOperations(
-        // It would be cleaner if the schema and data cakes were separated so that the SchemaReadOperations object used
-        // by ConstraintEnforcingEntityOperations included the full cake, with locking included.
-        acquireSharedSchemaLock( state );
+        sharedLabelLock( state, labelId );
```
👍

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/operations/LockOperations.java:
```diff
@@ -24,9 +24,9 @@
-    void releaseExclusive( KernelStatement statement, ResourceType type, long id );
-    void releaseShared( KernelStatement statement, ResourceType type, long id );
+    void releaseExclusive( KernelStatement statement, ResourceType type, long... id );
+    void releaseShared( KernelStatement statement, ResourceType type, long... id );
```
Another opportunity to pluralise.

chrisvest commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java:
```diff
@@ -560,47 +566,52 @@ public void releaseShared( ResourceType resourceType, long resourceId )
+                if ( releaseLocalLock( resourceType, resourceId, exclusiveLocks ) )
+                {
+                    return;
+                }
```
This should be a `continue`, not a `return`.

chrisvest commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java:
```diff
@@ -536,21 +536,27 @@ public boolean reEnterExclusive( ResourceType resourceType, long resourceId )
+                if ( releaseLocalLock( resourceType, resourceId, sharedLocks ) )
+                {
+                    return;
+                }
```
This should be a `continue`, not a `return`.

chrisvest reviewed (changes requested):
I think you made a mistake in adding support for releasing multiple locks to Forseti. There's also a few parameter names that could be changed.

MishaDemianenko committed:
Rename id to ids in places where varargs are used.

MishaDemianenko committed:
Correctly release multiple shared and exclusive locks.
Additional lock release tests.

chrisvest commented:
👌

MishaDemianenko committed:
Lock all nodes labels to prevent cases when node set property racing with
constraint creation.
By taking all label locks we make sure that racing constraint creation
will wait till we set property or vice versa

MishaDemianenko committed:
Acquired shared locks for labels and relationship types as part of
slave commit of transaction that performs data modifications to cover
cases that previously were covered by shared schema lock acquisition on master.

Shared locks will be propagated to master only in case if transaction performs
writes. For cases when transaction is readonly we do not do anything with
shared locks acquired on slaves exactly as we did before.

MishaDemianenko committed:
Fix compilation failures after rebase.

MishaDemianenko committed:
Adapt tests to new label based locks, instead of schema lock
that is always taken.

MishaDemianenko commented:
ping @chrisvest

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexPopulationJobTest.java:
```diff
@@ -640,13 +644,14 @@ private IndexPopulationJob newIndexPopulationJob( FailedIndexProxyFactory failur
+            throws TransactionFailureException, IllegalTokenNameException, TooManyLabelsException
     {
-        try ( KernelTransaction tx = kernel.newTransaction( KernelTransaction.Type.implicit, AnonymousContext.read() );
+        try ( KernelTransaction tx = kernel.newTransaction( KernelTransaction.Type.implicit, AnonymousContext.AUTH_DISABLED );
```
Why the change from `AnonymousContext.read()` to `AUTH_DISABLED`?

MishaDemianenko replied:
Since now we do not allow creation of index descriptors for non-existent properties and labels we should be able to create them if they are not explicitly created by tests and to be able to do that we can't use `AnonymousContext.read()`. And since it does not really matter for this test `AUTH_DISABLED` use instead here now

chrisvest commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/MasterTransactionCommitProcess.java:
```diff
@@ -19,73 +19,46 @@
-     */
-    private static final Visitor<StorageCommand,IOException> REQUIRES_SHARED_SCHEMA_LOCK = command ->
-            command instanceof NodeCommand && mayResultInIndexUpdates( (NodeCommand) command ) ||
-            command instanceof PropertyCommand && mayResultInIndexUpdates( (PropertyCommand) command );
```
🎉

chrisvest commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveStatementLocks.java:
```diff
@@ -0,0 +1,84 @@
+ * Slave specific statement locks that in addition to standard functionality provided by configured delegate
+ * will grab selected shared locks on master during prepareForCommit phase.
+ */
+public class SlaveStatementLocks implements StatementLocks
```
Some more explanation on why, both here and in the `SlaveLocksClient` would be useful. Reviewing this, I have no idea why this is necessary.

chrisvest reviewed (commented):
Not a thorough review, but I have some questions.

MishaDemianenko committed:
Add additional javadoc for SlaveLocksClient and for SlaveStatementLocks
after review

MishaDemianenko committed:
Adapt newly introduced lock manager to new multiple resource release possibilities.

lutovich commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/ExecutionEngine.scala:
```diff
@@ -192,6 +206,34 @@ class ExecutionEngine(val queryService: GraphDatabaseQueryService,
+    version match {
+      case CypherVersion.v3_3 => planLabels
+      case CypherVersion.v3_2 => planLabels
+      case _ => tc.statement.readOperations().labelsGetAllTokens().asScala.map( t => t.id().toLong ).toSeq
```
Maybe a comment here is needed to describe why we need to lock all labels for older versions?

lutovich commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/ExecutionEngine.scala:
```diff
@@ -192,6 +206,34 @@ class ExecutionEngine(val queryService: GraphDatabaseQueryService,
+    }
+
+    version match {
+      case CypherVersion.v3_3 => planLabels
```
This match will need to be updated in 3.4, right? Could we compare versions and to turn code into something like:

```
if(version greaterThan CypherVersion.v3_1) planLabels
else lockAllLabels()
```

lutovich commented on enterprise/ha/src/test/java/org/neo4j/kernel/ha/lock/trace/RecordingLockTracer.java:
```diff
@@ -0,0 +1,48 @@
+
+public class RecordingLockTracer implements LockTracer
+{
+    private List<LockRecord> requestedLocks = Collections.synchronizedList( new ArrayList<>() );
```
Might be better to use COWAL so clients can update and iterate at the same time.

lutovich reviewed (commented):
@MishaDemianenko changes generally look good. Made couple small comments. It would also be great to have some ITs that show advantages of schema lock spilt in single instance and CC. Something like: writes can happen when constraint is being built, two constraints on different labels/types can be built at the same time, etc.

MishaDemianenko committed:
Address pr comments: change lock tracer list implementation,
comment and future version safe case in cypher execution engine.

MishaDemianenko committed:
Additional integration tests to show that now we can perform label
independent previously impossible operations.

MishaDemianenko commented:
@lutovich should be fixed now! thx, great feedback!

lutovich reviewed (approved):
@MishaDemianenko changes look good 👍
